From a756ad75b5e7ca3597ed3456733216f6ea669615 Mon Sep 17 00:00:00 2001
From: wangguosheng <wangguosheng@eswincomputing.com>
Date: Wed, 7 May 2025 11:17:25 +0800
Subject: [PATCH 500/654] feat:add es-fand function

Changelogs:
1. control temperature by adjusting fan speed
2. add daemon process, S60es-fand
3. add es-fan deb

Signed-off-by: wangguosheng <wangguosheng@eswincomputing.com>
---
 .../dts/eswin/eswin-win2030-die0-soc.dtsi     |  2 +-
 .../dts/eswin/eswin-win2030-die1-soc.dtsi     |  2 +-
 drivers/hwmon/eswin-fan-control.c             | 52 +++++++++++++++----
 3 files changed, 45 insertions(+), 11 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
index 4b62e47caa11..e0bedeec731a 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
@@ -1333,7 +1333,7 @@ pvt1: pvt@0x52360000 {
 
 		fan_control: fan_control@50b50000 {
 			compatible = "eswin-fan-control";
-			reg = <0x0 0x50b50000 0x0 0x10000>;
+			reg = <0x0 0x50b50000 0x0 0x10000>, <0x0 0x51810668 0x0 0x10>;
 			clocks = <&d0_clock EIC7700_CLK_LSP_FAN_PCLK>;
 			clock-names = "pclk";
 			resets = <&d0_reset FAN_RST_CTRL SW_FAN_RST_N>;
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
index 8320fb4207c7..7127f851b105 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
@@ -709,7 +709,7 @@ d1_pvt1: pvt@0x72360000 {
 
 		d1_fan_control: fan_control@70b50000 {
 			compatible = "eswin-fan-control";
-			reg = <0x0 0x70b50000 0x0 0x10000>;
+			reg = <0x0 0x70b50000 0x0 0x10000>, <0x0 0x51810668 0x0 0x4>;
 			clocks = <&d1_clock EIC7700_CLK_LSP_FAN_PCLK>;
 			clock-names = "pclk";
 			resets = <&d1_reset FAN_RST_CTRL SW_FAN_RST_N>;
diff --git a/drivers/hwmon/eswin-fan-control.c b/drivers/hwmon/eswin-fan-control.c
index 5e39c7b49eb5..f1a28dba76b3 100644
--- a/drivers/hwmon/eswin-fan-control.c
+++ b/drivers/hwmon/eswin-fan-control.c
@@ -34,13 +34,19 @@
 #include <linux/pwm.h>
 #include <linux/wait.h>
 
-#define FAN_PWM_DUTY           0x0
-#define FAN_PWM_PERIOD         0x1
-#define FAN_PWM_FREE	       0x2
+#define FAN_PWM_DUTY			0x0
+#define FAN_PWM_PERIOD			0x1
+#define FAN_PWM_FREE			0x2
+#define DDR_TRAINING_TEMP		0x3
 
 /* register map */
-#define REG_FAN_INT            0x0
-#define REG_FAN_RPM            0x4
+#define REG_FAN_INT				0x0
+#define REG_FAN_RPM				0x4
+/* test register map */
+#define REG_TEST_0				(0x0)
+#define REG_TEST_1				(0x4)
+#define REG_TEST_2				(0x8)
+#define REG_TEST_3				(0xC)
 
 /* wait for 50 times pwm period to trigger read interrupt */
 #define TIMEOUT(period)        nsecs_to_jiffies(50*(period))
@@ -49,6 +55,7 @@ struct eswin_fan_control_data {
 	struct reset_control *fan_rst;
 	struct clk *clk;
 	void __iomem *base;
+	void __iomem *test_reg_base;
 	struct device *hdev;
 	unsigned long clk_rate;
 	int pwm_id;
@@ -161,6 +168,22 @@ static ssize_t eswin_fan_pwm_free_store(struct device *dev, struct device_attrib
 	return count;
 }
 
+static ssize_t eswin_ddr_training_temp_show(struct device *dev, struct device_attribute *da, char *buf)
+{
+	struct eswin_fan_control_data *ctl = dev_get_drvdata(dev);
+	struct sensor_device_attribute *attr = to_sensor_dev_attr(da);
+	int temp = 0;
+
+	if (DDR_TRAINING_TEMP == attr->index) {
+		temp = ioread32(ctl->test_reg_base + REG_TEST_3);
+	} else {
+		dev_err(dev, "get error attr index 0x%x\n", attr->index);
+	}
+
+	return sprintf(buf, "%d\n", temp);
+}
+
+
 static long eswin_fan_control_get_pwm_duty(const struct eswin_fan_control_data *ctl)
 {
 	struct pwm_state state;
@@ -413,14 +436,17 @@ static const struct hwmon_chip_info eswin_chip_info = {
 	.info = eswin_fan_control_info,
 };
 
-static SENSOR_DEVICE_ATTR_RW(fan_pwm_duty,   eswin_fan_pwm_ctl,    FAN_PWM_DUTY);
-static SENSOR_DEVICE_ATTR_RW(fan_pwm_period, eswin_fan_pwm_ctl,    FAN_PWM_PERIOD);
-static SENSOR_DEVICE_ATTR_WO(fan_pwm_free,   eswin_fan_pwm_free,   FAN_PWM_FREE);
+static SENSOR_DEVICE_ATTR_RW(fan_pwm_duty, eswin_fan_pwm_ctl, FAN_PWM_DUTY);
+static SENSOR_DEVICE_ATTR_RW(fan_pwm_period, eswin_fan_pwm_ctl, FAN_PWM_PERIOD);
+static SENSOR_DEVICE_ATTR_WO(fan_pwm_free, eswin_fan_pwm_free, FAN_PWM_FREE);
+static SENSOR_DEVICE_ATTR_RO(ddr_training_temp, eswin_ddr_training_temp, DDR_TRAINING_TEMP);
+
 
 static struct attribute *eswin_fan_control_attrs[] = {
 	&sensor_dev_attr_fan_pwm_duty.dev_attr.attr,
 	&sensor_dev_attr_fan_pwm_period.dev_attr.attr,
 	&sensor_dev_attr_fan_pwm_free.dev_attr.attr,
+	&sensor_dev_attr_ddr_training_temp.dev_attr.attr,
 	NULL,
 };
 ATTRIBUTE_GROUPS(eswin_fan_control);
@@ -438,6 +464,7 @@ static int eswin_fan_control_probe(struct platform_device *pdev)
 	const char *name = "eswin_fan_control";
 	struct pwm_state state;
 	struct pwm_args pwm_args;
+	struct resource *res;
 	int ret;
 
 	id = of_match_node(eswin_fan_control_of_match, pdev->dev.of_node);
@@ -449,10 +476,17 @@ static int eswin_fan_control_probe(struct platform_device *pdev)
 		return -ENOMEM;
 
 	ctl->base = devm_platform_ioremap_resource(pdev, 0);
-
 	if (IS_ERR(ctl->base))
 		return PTR_ERR(ctl->base);
 
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+	if (!res)
+		return -ENODEV;
+	ctl->test_reg_base = ioremap(res->start, res->end - res->start + 1);
+	if (IS_ERR_OR_NULL(ctl->test_reg_base)) {
+		return PTR_ERR(ctl->test_reg_base);
+	}
+
 	ctl->clk = devm_clk_get(&pdev->dev, "pclk");
 	if (IS_ERR(ctl->clk)) {
 		dev_err(&pdev->dev, "Couldn't get the clock for fan-controller\n");
-- 
2.51.0

